### PR TITLE
RPInitiatedLogoutRequest fixes

### PR DESCRIPTION
**Pull request by cblakeley**
This patch relates to:
  * https://github.com/solid/oidc-op/issues/10#issuecomment-426233886
  * https://github.com/solid/oidc-op/issues/10#issuecomment-426602431
The fixes were needed to get RPInitiatedLogoutRequest working.

### CODE REVIEW

**Comment by dmitrizagidulin** [comment]
@cblakeley - can you explain a little more what this PR is trying to do? It's not clear from the comments to PR #10.

File: src/handlers/RPInitiatedLogoutRequest.js [hunk 1]
```diff
@@ -40,20 +41,17 @@ class RPInitiatedLogoutRequest extends BaseRequest {
...
 
     return Promise
       .resolve(request)
       .then(request.validate)
-
       // From: https://openid.net/specs/openid-connect-session-1_0.html#RPLogout
       // At the logout endpoint, the OP SHOULD ask the End-User whether they want
       // to log out of the OP as well. If the End-User says "yes", then the OP
       // MUST log out the End-User.
-      .then(host.logout)
```
Comment (dmitrizagidulin) [inline src/handlers/RPInitiatedLogoutRequest.js]: Why was the `host.logout` callback removed? That's a key part of the mechanism.

File: src/handlers/RPInitiatedLogoutRequest.js [hunk 2]
```diff
@@ -40,20 +41,17 @@ class RPInitiatedLogoutRequest extends BaseRequest {
...
-
-      .then(request.redirectToGoodbye.bind(request))
+      .then(request.clearUserSession.bind(request))
```
Comment (dmitrizagidulin) [inline src/handlers/RPInitiatedLogoutRequest.js]: Ah, ok, so it seems like you're shifting clearing the user session from the `host.logout` callback to `clearUserSession` on this request. However, that's not the right place to do it - this request has no idea of what session or cookie mechanism the host provider is using. That's why it's letting it do its own session clearing in `host.logout(request)`.

**Comment by cblakeley** [comment]
@dimitrizagidulin: Some explanation... I think there are two OIDC logout routines for Solid's OP:

* logout v1: oidc-auth-manager --> host_api.js#logout --> LogoutRequest#handle
* logout v2: oidc-op --> Provider#logout --> RPInitiatedLogoutRequest#handle

I gather logout v2 differs from logout v1 in basically one respect - it validates that any post_logout_redirect_uri, supplied by the RP with an RP-initiated logout request, was pre-registered with the OP. So logout v2 tightens up the OP's adherence to the OIDC spec. logout v1 doesn't check the supplied post_logout_redirect_uri.

At present, Solid uses logout v1, logout v2 isn't being used. This PR fixes logout v2 and https://github.com/solid/node-solid-server/pull/857 modifies Solid to use it. Without https://github.com/solid/node-solid-server/pull/857, /logout appears to bind directly to LogoutRequest#handle, so RPInitiatedLogoutRequest#handle wasn't being invoked.

A call to JWKSet.importKeys was needed to set cryptoKey, otherwise it wasn't set and validation of the id_token_hint failed.

File: src/handlers/RPInitiatedLogoutRequest.js (hunk 1 above)
Comment (dmitrizagidulin) [reply src/handlers/RPInitiatedLogoutRequest.js]: Ok, got it. Let me look around the repos some more, maybe see if there's a clearer way..

**Comment by dmitrizagidulin** [comment]
@cblakeley Great work on all of this in general, I apologize for any confusion. :) This is one of the hardest subsystems we have in Solid, and it certainly doesn't help that the logic is divided into like a half dozen repos.

**Comment by dmitrizagidulin** [comment]
(still reviewing, stand by.)

**Comment by cblakeley** [comment]
Can this PR please be reviewed, so the fix can be finalized? OpenLink would like to get this revised RP-initiated logout code (which conforms more closely to the spec) into our OPL forks. Because the Solid team may change the implementation of this fix, we'd rather not merge the current implementation into our forks.
We'd prefer to hold off until the finalized version is committed to the master Solid repos first, but we've been waiting a while now. (Work on logout related issues started back in June - we'd like to get this whole logout issue finished). 

Impact - OIDC clients will need to be modified to pre-register post_logout_redirect_uris. Any query string parameters currently included with the post_logout_redirect_uri may need to be encoded in the state parameter instead.

**Comment by kidehen** [comment]
@kjetilk ,

Please note this item, which relates to your pending release.

/cc @RubenVerborgh @justinwb @dmitrizagidulin

**Comment by RubenVerborgh** [comment]
@cblakeley Is there any way you can explain for non-OIDC-experts what this is doing? Something along the lines of "before this PR, xyz, after this PR, abc".

The difficulty is that we don't know what "RPInitiatedLogoutRequest fixes" precisely means. What is the issue being fixed? I get that https://github.com/solid/oidc-op/pull/11#issuecomment-429031969 is an explanation, but some basic things aren't answered for me. Why is logout route v2 needed? (How) is it better than v1? Should v1 be dropped?

**Comment by cblakeley** [comment]
@RubenVerborgh @dmitrizagidulin /cc @kidehen: 

Before this PR:

* RP initiated logout requests were handled by LogoutRequest#handle (logout v1)
* logout v1 does the job (insofar as the RP session with the OP is destroyed) but it doesn't fully conform to the OIDC spec (https://openid.net/specs/openid-connect-session-1_0.html#RPLogout).
  * It doesn't require RPs to pre-register post-logout redirect URIs.
  * It doesn't validate any `post_logout_redirect_uri` supplied by the RP, by checking it was pre-registered.
  * The `post_logout_redirect_uri` could include arbitrary query string parameters (e.g. for maintaining state), whereas the spec only allows 3 specific parameters.
* Dmitri addressed the above conformance issues by implementing RPInitiatedLogoutRequest#handle (logout v2), but I found problems with RPInitiatedLogoutRequest:
  * The new code wasn't being invoked - the route binding for /logout appeared to me to be still bound to logout v1, not logout v2.
  * Some other bugs in RPInitiatedLogoutRequest

After this PR:

* logout v2 is bound to /logout and is invoked.
* The bugs in the implementation are fixed.
* The post logout redirect URI handling conforms to the spec.
  * The only query parameters allowed with the logout request are: `post_logout_redirect_uri`, `id_token_hint` and `state`.

Impact:

As mentioned previously above, OIDC clients will need to be modified to pre-register `post_logout_redirect_uris` as part of the OIDC RP registration. Any query string parameters currently included with the `post_logout_redirect_uri` may need to be encoded in the state parameter instead.

@dmitrizagidulin commented above that he thought some of my changes in this PR would need refactoring. I'm hoping the core Solid Team can address this. 

HTH

**Comment by RubenVerborgh** [comment]
Thanks a lot @cblakely, I indeed hope this is sufficient for the team.

**Comment by dmitrizagidulin** [comment]
👍 from me on this PR.

**Pull request merged**